pkg/sql/create_stats.go

```diff
@@ -742,6 +767,21 @@ func checkRunningJobs(ctx context.Context, job *jobs.Job, p JobExecContext) erro
 	return nil
 }
 
+// checkJobRunningForTooLong checks whether the given job, which was created at
+// the given time, has been running for too long. If it has, mark it as failed.
+func checkJobRunningForTooLong(
+	ctx context.Context, txn isql.Txn, jobID jobspb.JobID, created time.Time, p JobExecContext,
+) (canceled bool, err error) {
+	// Check that the job was not created more than maxStatsJobDuration ago.
+	if created.Add(maxStatsJobDuration.Get(p.ExecCfg().SV())).After(timeutil.Now()) {
+		return false, nil
+	}
+
+	// The job has been running for too long, so mark it as failed.
+	err = p.ExecCfg().JobRegistry.Failed(ctx, txn, jobID, errors.Newf("stats job has been running for too long"))
```

Comment:
I'm not sure this method should actually be exported at all; this is moving the job directly into the failed state even while it is still running, whereas callers should be requesting cancellation, which would move it to the cancel requested state, and then when the job noticed the cancelation request and actually exits resume it would move into cancelling, and then when onFailOrCancel completes it would move to cancelled. Essentially the job system relies on cooperative scheduling, where the caller requests a state and then the job moves to it, rather than the caller moving the job.  Moving the job directly from running to failed may well violate some job system assumptions so I'm a little wary of this change. We also _just_ had an issue due to stats jobs directly messing with the jobs system in bespoke ways (partially deleting the aborted jobs) so I'm also a little paranoid about this today.

(this is why I was suggesting a simpler approach of simply having the resumer limit its own execution time, rather than trying to reach into one job from another and change its state).